ext/drx/drx_buf.c

```diff
@@ -648,6 +649,96 @@ drx_buf_insert_buf_store(void *drcontext, drx_buf_t *buf, instrlist_t *ilist,
     }
 }
 
+static void
+insert_load(void *drcontext, instrlist_t *ilist, instr_t *where,
+                    reg_id_t dst, reg_id_t src, opnd_size_t opsz)
+{
+    switch (opsz) {
+    case OPSZ_1:
+        MINSERT(ilist, where, XINST_CREATE_load_1byte
+                (drcontext,
+                 opnd_create_reg(reg_resize_to_opsz(dst, opsz)),
+                 opnd_create_base_disp(src, DR_REG_NULL, 0, 0, opsz)));
+        break;
+    case OPSZ_2:
+        MINSERT(ilist, where, XINST_CREATE_load_2bytes
+                (drcontext,
+                 opnd_create_reg(reg_resize_to_opsz(dst, opsz)),
+                 opnd_create_base_disp(src, DR_REG_NULL, 0, 0, opsz)));
+        break;
+    case OPSZ_4:
+#if defined(X86_64) || defined(AARCH64)
+    case OPSZ_8:
+#endif
+        MINSERT(ilist, where, XINST_CREATE_load
+                (drcontext,
+                 opnd_create_reg(reg_resize_to_opsz(dst, opsz)),
+                 opnd_create_base_disp(src, DR_REG_NULL, 0, 0, opsz)));
+        break;
+    default:
+        DR_ASSERT(false);
+        break;
+    }
+}
+
+DR_EXPORT
+void
+drx_buf_insert_buf_memcpy(void *drcontext, drx_buf_t *buf, instrlist_t *ilist,
+                          instr_t *where, reg_id_t dst, reg_id_t src, ushort len)
+{
+    opnd_size_t opsz = opnd_size_from_bytes(len);
+
+    if (len > sizeof(app_pc)) {
+        /* slow path, we call into our personal memcpy() implementation */
+        /* XXX: We should elect to write our own memcpy() built in DR IR. This
+         * would potentially be slower than the glibc memcpy, but then we can
+         * incrementally update the buffer pointer so the user can recover
+         * from partial writes.
+         * XXX: It's also unlikely that if glibc's memcpy faults, we'd be able
+         * to recover from it -- an in-app fault will not be handled by us, and
+         * also the buffer pointer is not updated incrementally.
```

Comment:
First, I thought there was a problem with a fault in memcpy even if it's not in the client lib: you said your drx_buf fault handling code assumed the faulting store instruction had a certain form, right?  And that you expected fault handling to fail, independently of #50.  This is why I thought this was not yet complete for all cases, not just memcpy inlined into the client.

Re: DR IR vs direct asm code + #50: direct code seems easier to read and maintain?  But then you have to implement #50 (for which we prob want a separate event for the unstable-client reason so that only clients needing it and willing to accept the tradeoff will bypass error reports).  Either way: for long term asm code + #50 seems the way to go but for getting it done quickly IR would be faster to imlement.